src/Markup/Avalonia.Markup/Data/Binding.cs

```diff
@@ -105,37 +105,53 @@ public Binding(string path, BindingMode mode = BindingMode.Default)
             
             ExpressionObserver observer;
 
+            var (node, mode)  = ExpressionObserverBuilder.Parse(Path, enableDataValidation, TypeResolver);
+
             if (ElementName != null)
             {
                 observer = CreateElementObserver(
                     (target as IStyledElement) ?? (anchor as IStyledElement),
                     ElementName,
-                    Path,
-                    enableDataValidation);
+                    node);
             }
             else if (Source != null)
             {
-                observer = CreateSourceObserver(Source, Path, enableDataValidation);
+                observer = CreateSourceObserver(Source, node);
             }
-            else if (RelativeSource == null || RelativeSource.Mode == RelativeSourceMode.DataContext)
+            else if (RelativeSource == null)
+            {
+                if (mode == SourceMode.Data)
```

Comment:
There seems to be duplicated code in this block. Could we do something like:

```
var relativeSource = RelativeSource ?? SourceModeToRelativeSource(mode);
```

And then use `relativeSource` to decide the source?

Reply:
I'd have to surface more data up from the expression parsing instead of just "bind to the data"/"bind to the control"

Reply:
Which would make the code a little messier I think.

Reply:
Ok, sure. Like I say, I didn't actually try it ;)